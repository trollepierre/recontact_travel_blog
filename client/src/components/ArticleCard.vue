<template>
  <div class="article-card">
    <article class="article">
      <header class="article__header">
        <a :href="articleUrl" @click.prevent.once="viewArticle">
          <h2 class="article__title">{{ articleTitle }}</h2>
        </a>
      </header>
      <div class="article__content">
        <img class="article__image" :src="article.imgLink" width="200"/>
      </div>
      <footer class="article__footer">
        <template v-if="adminMode">
          <button class="article__update-button" :disabled="isUpdateClicked"
                  @click.prevent.once="updateArticle">
            {{ $t("repairArticle") }}
          </button>
        </template>
        <template v-else>
          <button class="article__view-button"
                  @click.prevent.once="viewArticle">
            {{ $t("goToArticle") }}
          </button>
          <a :href="article.galleryLink" target="_blank" class="article__dropbox">
            <button class="article__dropbox-button">
              {{ $t("viewGallery") }}
            </button>
          </a>
        </template>
      </footer>
    </article>
  </div>
</template>

<script>
  import articlesApi from '@/api/articles';
  import notificationsService from '@/services/notifications';
  import translationsService from '@/services/translations';

  export default {
    name: 'ArticleCard',
    props: ['article', 'adminMode'],
    data() {
      return {
        isUpdateClicked: false,
      };
    },
    computed: {
      articleUrl() {
        return `/articles/${this.article.dropboxId}`;
      },
      articleTitle() {
        return translationsService.getTitle(this.article);
      },
    },
    methods: {
      viewArticle() {
        this.goToArticle();
      },

      updateArticle() {
        this.disableDeleteButton();
        notificationsService.information(this, this.$t('syncLaunched'));
        articlesApi.update(this.article.dropboxId)
          .then(() => {
            notificationsService.removeInformation(this);
            notificationsService.success(this, this.$t('syncDone'));
          })
          .then(() => this.goToArticle())
          .catch((err) => {
            notificationsService.removeInformation(this);
            notificationsService.error(this, `${this.$t('syncError')} ${err}`);
          });
      },

      disableDeleteButton() {
        this.isUpdateClicked = true;
      },

      goToArticle() {
        this.$router.push(this.articleUrl);
      },
    },

    i18n: {
      messages: {
        fr: {
          repairArticle: 'Réparer l‘article',
          goToArticle: 'Voir l‘article',
          viewGallery: 'Voir les photos',
          syncLaunched: 'La synchronisation est lancée ! Patientez quelques secondes...',
          syncDone: 'La synchronisation s‘est effectuée sans problème !',
          syncError: 'Erreur : Problème durant la synchronisation :',
        },
        en: {
          repairArticle: 'Repair the article',
          goToArticle: 'Read the article',
          viewGallery: 'Discover the pictures',
          syncLaunched: 'The synchronisation is launched! Please wait...',
          syncDone: 'The synchronisation succeeds!',
          syncError: 'Error during the synchronisation:',
        },
      },
    },
  };
</script>

<style scoped>
  h2 {
    font-weight: normal;
  }

  img {
    max-width: 100%;
    max-height: 100%;
  }

  .sr-only {
    position: absolute;
    width: 1px;
    height: 1px;
    padding: 0;
    margin: -1px;
    overflow: hidden;
    clip: rect(0, 0, 0, 0);
    border: 0;
    display: block;
  }

  .article {
    min-width: 260px;
    max-width: 260px;
    background: #ffffff;
    border-radius: 4px !important;
    box-shadow: 0 1px 1px rgba(0, 0, 0, .15);
    border: 1px solid rgba(0, 0, 0, .09);
    display: flex;
    flex-direction: column;
    color: #535a60;
  }

  .article__header {
    border-bottom: 1px solid #e6e6e6;
    height: 34px;
    padding: 15px;
    display: flex;
    justify-content: space-between;
  }

  .article__header a {
    margin: auto;
  }

  .article__title {
    font-size: 16px;
    font-weight: 700;
    line-height: 17px;
    color: #07c;
    margin: auto;
    overflow: hidden;
    text-overflow: ellipsis;
    max-width: 200px;
  }

  .article__content {
    font-size: 15px;
    padding: 15px;
    height: 150px;
    color: #000;
    text-align: center;
    display: flex;
    align-content: center;
    justify-content: center;
  }

  .article__content > p {
    margin-top: 0;
  }

  .article__footer {
    text-align: center;
    padding: 15px;
    border-top: 1px solid #e6e6e6;
  }

  .article__footer button {
    text-transform: uppercase;
    color: #d14800;
    background: #ffffff;
    border: 1px solid #d14800;
    cursor: pointer;
    padding: 15px 30px;
    border-radius: 4px;
    width: 100%;
    margin-bottom: 10px;
    font-weight: 700;
  }

  .article__footer button:hover {
    background: #d14800;
    color: #ffffff;
  }

  .article__footer button:disabled,
  .article__footer button:active {
    background: #BDBDBD;
    border-color: #616161;
    color: #FAFAFA;
    cursor: auto;
  }

  @media only screen and (max-width: 640px) {
    .article {
      max-width: inherit;
    }
  }
</style>
