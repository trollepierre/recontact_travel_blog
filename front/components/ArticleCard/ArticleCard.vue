<template>
  <div class="article-card">
    <article class="article">
      <header class="article__header">
        <a
          :href="articleUrl"
          @click.prevent.once="viewArticle">
          <h2 class="article__title">
            {{ articleTitle }}
          </h2>
        </a>
      </header>
      <div
        class="article__content"
        @click.prevent.once="viewArticle">
        <img
          :src="article.imgLink"
          :alt="articleTitle"
          class="article__image"
          width="200">
      </div>
      <footer class="article__footer">
        <template v-if="adminMode">
          <button
            :disabled="isUpdateClicked"
            class="article__update-button"
            @click.prevent.once="updateArticle">
            {{ $t("repairArticle") }}
          </button>
          <button
            :disabled="isDeleteClicked"
            class="article__delete-button article__footer_hidden"
            @click.prevent.once="deleteArticle">
            {{ $t("deleteArticle") }}
          </button>
        </template>
        <template v-else>
          <button
            class="article__view-button"
            @click.prevent.once="viewArticle">
            {{ $t("goToArticle") }}
          </button>
          <a
            :href="article.galleryLink"
            target="_blank"
            class="article__dropbox">
            <button class="article__dropbox-button">
              {{ $t("viewGallery") }}
            </button>
          </a>
        </template>
      </footer>
    </article>
  </div>
</template>

<script>
  import articlesApi from '../../services/api/articles'
  import notificationsService from '../../services/services/notifications'
  import translationsService from '../../services/services/translations'

  export default {
    name: 'ArticleCard',
    props: {
      adminMode: { type: Boolean, default: () => false },
      article: { type: Object, default: () => {} },
    },
    data() {
      return {
        isUpdateClicked: false,
        isDeleteClicked: false,
      }
    },
    computed: {
      articleUrl() {
        return `/articles/${this.article.dropboxId}`
      },
      articleTitle() {
        const language = this.$store.state.locale
        return translationsService.getTitle(this.article, language)
      },
    },
    methods: {
      viewArticle() {
        this.goToArticle()
      },

      updateArticle() {
        this.trackEvent()
        this.disableUpdateButton()
        notificationsService.information(this, this.$t('syncLaunched'))
        articlesApi.update(this.article.dropboxId)
          .then(() => {
            notificationsService.removeInformation(this)
            notificationsService.success(this, this.$t('syncDone'))
          })
          .then(() => this.goToArticle())
          .catch(err => {
            notificationsService.removeInformation(this)
            notificationsService.error(this, `${this.$t('syncError')} ${err}`)
          })
      },

      disableUpdateButton() {
        this.isUpdateClicked = true
      },

      trackEvent() {
        this.$ga.event({
          eventCategory: 'Article Card',
          eventAction: 'update',
          eventLabel: `article ${this.article.dropboxId} is updated`,
        })
      },

      deleteArticle() {
        this.disableDeleteButton()
        notificationsService.information(this, this.$t('deleteLaunched'))
        articlesApi.delete(this.article.dropboxId)
          .then(() => {
            notificationsService.removeInformation(this)
            notificationsService.success(this, this.$t('deleteDone'))
          })
          .catch(err => {
            notificationsService.removeInformation(this)
            notificationsService.error(this, `${this.$t('deleteError')} ${err}`)
          })
      },

      disableDeleteButton() {
        this.isDeleteClicked = true
      },

      goToArticle() {
        this.$router.push(this.articleUrl)
      },
    },

    i18n: {
      messages: {
        fr: {
          repairArticle: 'Réparer l’article',
          deleteArticle: 'Supprimer l’article',
          goToArticle: 'Voir l’article',
          viewGallery: 'Voir les photos',
          syncLaunched: 'La synchronisation est lancée ! Patientez quelques secondes...',
          syncDone: 'La synchronisation s’est effectuée sans problème  !',
          syncError: 'Erreur : Problème durant la synchronisation  :',
          deleteLaunched: 'La suppression est lancée ! Patientez quelques secondes...',
          deleteDone: 'La suppression s’est effectuée sans problème  !',
          deleteError: 'Erreur : Problème durant la suppression  :',
        },
        en: {
          repairArticle: 'Repair the article',
          deleteArticle: 'Delete the article',
          goToArticle: 'Read the article',
          viewGallery: 'Discover the pictures',
          syncLaunched: 'The synchronisation is launched! Please wait...',
          syncDone: 'The synchronisation succeeds!',
          syncError: 'Error during the synchronisation:',
          deleteLaunched: 'The deletion is launched! Please wait...',
          deleteDone: 'The deletion succeeds!',
          deleteError: 'Error during the deletion:',
        },
      },
    },
  }
</script>

<style scoped>
  h2 {
    font-weight: normal;
  }

  img.article__image {
    max-width: 100%;
    max-height: 40vw;
  }

  @media only screen and (max-width: 640px) {
    img.article__image {
      max-height: 100%;
    }
  }

  .sr-only {
    position: absolute;
    width: 1px;
    height: 1px;
    padding: 0;
    margin: -1px;
    overflow: hidden;
    clip: rect(0, 0, 0, 0);
    border: 0;
    display: block;
  }

  .article {
    min-width: 260px;
    max-width: 260px;
    background: #ffffff;
    border-radius: 4px !important;
    box-shadow: 0 1px 1px rgba(0, 0, 0, .15);
    border: 1px solid rgba(0, 0, 0, .09);
    display: flex;
    flex-direction: column;
    color: #535a60;
  }

  .article__header {
    border-bottom: 1px solid #e6e6e6;
    height: 34px;
    padding: 15px;
    display: flex;
    justify-content: space-between;
  }

  .article__header a {
    margin: auto;
  }

  .article__title {
    font-size: 16px;
    font-weight: 700;
    line-height: 17px;
    color: #07c;
    margin: auto;
    overflow: hidden;
    text-overflow: ellipsis;
    max-width: 200px;
  }

  .article__content {
    font-size: 15px;
    padding: 15px;
    height: 150px;
    color: #000;
    text-align: center;
    display: flex;
    align-content: center;
    justify-content: center;
  }

  .article__content > p {
    margin-top: 0;
  }

  .article__footer {
    text-align: center;
    padding: 15px;
    border-top: 1px solid #e6e6e6;
  }

  .article__footer button {
    text-transform: uppercase;
    color: #d14800;
    background: #ffffff;
    border: 1px solid #d14800;
    cursor: pointer;
    padding: 15px 30px;
    border-radius: 4px;
    width: 100%;
    margin-bottom: 10px;
    font-weight: 700;
  }

  .article__footer button.article__footer_hidden {
    color: #f7b5a9;
  }

  .article__footer button:hover {
    background: #d14800;
    color: #ffffff;
  }

  .article__footer button:disabled,
  .article__footer button:active {
    background: #BDBDBD;
    border-color: #616161;
    color: #FAFAFA;
    cursor: auto;
  }

  @media only screen and (max-width: 640px) {
    .article {
      max-width: inherit;
    }
  }
</style>
