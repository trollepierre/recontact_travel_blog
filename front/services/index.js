export { isEmptyPlus } from './ramda/ramda-utils'
export { isWww, isCecile } from './location/location'
export { iframeDimensions } from './utils/iframe/iframe-utils'
export { sortByDropboxId } from './utils/article/articlesSorter'
