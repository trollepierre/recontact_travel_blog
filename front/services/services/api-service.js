import axios from 'axios'
import { prop } from 'ramda'
import env from '../env/env'
import logger from './logger-service'

const apiPath = `${env('API_URL') }api/`

const getAll = async path => {
  try {
    const url = `${apiPath}${path}`
    const response = await axios.get(url, { json: true })
    return prop('data', response)
  } catch (error) {
    logger.error(error.message)
    throw error
  }
}

const post = async (path, data) => {
  try {
    const url = `${apiPath}${path}`
    const response = await axios.post(url, { ...data, json: true })
    return prop('data', response)
  } catch (error) {
    logger.error(error.message)
    throw error
  }
}

const put = async (path, data) => {
  try {
    const url = `${apiPath}${path}`
    const response = await axios.patch(url, { ...data, json: true })
    return prop('data', response)
  } catch (error) {
    logger.error(error.message)
    throw error
  }
}

const deleteById = async path => {
  try {
    const url = `${apiPath}${path}`
    const response = await axios.delete(url)
    return prop('data', response)
  } catch (error) {
    logger.error(error.message)
    throw error
  }
}

export default {
  get: getAll,
  post,
  put,
  delete: deleteById,
}
