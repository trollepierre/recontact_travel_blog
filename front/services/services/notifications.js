export default {
  information(message) {
    console.log(message)
  },

  error(message) {
    console.error(message)
    window.alert(message) // eslint-disable-line no-alert
  },

  warn(message) {
    console.warn(message)
  },
}
