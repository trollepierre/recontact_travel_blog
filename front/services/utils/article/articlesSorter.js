const getDropboxId = article => parseInt(article.dropboxId, 10)

const sortByDropboxId = articles => articles.sort(
  (articleA, articleB) => getDropboxId(articleA) < getDropboxId(articleB) ? 1 : -1,
)

export {
  sortByDropboxId,
}
