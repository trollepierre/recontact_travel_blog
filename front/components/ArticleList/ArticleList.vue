<template>
  <div class="article-list">
    <section class="article-results">
      <h1 class="article-results__title hidden">
        {{ hiddenTitle }}
      </h1>
      <p class="article-results__title">
        {{ title }}
      </p>
      <p class="article-results__title h2">
        {{ subtitle }}
        <span
          v-if="!isCecile"
          class="article-results__title h3">
          {{ lastPosition }}
        </span>
      </p>
      <template v-if="adminMode">
        <admin-dashboard @updateLastPositionData="updateLastPositionData"/>
      </template>
      <ul class="article-results__list">
        <li
          v-for="article in articles"
          :key="article.dropboxId"
          class="article-results__item">
          <article-card
            :article="article"
            :lazy="isLazyArticle(article.dropboxId)"
            :admin-mode="adminMode"/>
        </li>
      </ul>
    </section>
  </div>
</template>

<script>
  import ArticleCard from '../ArticleCard/ArticleCard.vue'
  import AdminDashboard from '../AdminDashboard/AdminDashboard.vue'
  import articlesApi from '../../services/api/articles'
  import positionsApi from '../../services/api/positions'
  import { isCecile, sortByDropboxId } from '../../services'
  import translationService from '../../services/services/translations'
  import { IS_DESKTOP } from '../../services/utils/responsive/responsive-utils'

  export default {
    name: 'ArticleList',
    components: {
      AdminDashboard,
      ArticleCard,
    },
    props: {
      adminMode: { type: Boolean, default: () => false },
      articlesNumberLimit: { type: Number, default: () => 0 },
    },
    data() {
      return {
        articles: [],
        lastPosition: '',
      }
    },
    computed: {
      hiddenTitle() {
        return this.adminMode ? this.$t('fixWebsite') : this.$t('theArticlesOfTheTrip')
      },
      title() {
        return isCecile() ? 'Mon petit Cadeau de Saint Valentin' : this.$t('title')
      },
      subtitle() {
        return isCecile() ? 'recharge le site si jamais l’article ne s’est pas chargé' : this.$t('lastKnownPosition')
      },
      isCecile() {
        return isCecile()
      },
    },
    mounted() {
      this.getArticles()
      this.getLastPosition()
    },
    methods: {
      getArticles() {
        articlesApi.fetchAll(this.articlesNumberLimit)
          .then(articles => {
            this.articles = sortByDropboxId(articles)
          })
      },

      updateLastPositionData({
        place, time, placeEn, timeEn,
      }) {
        const language = this.$store.state.locale
        if (translationService.isFrancophone(language)) {
          this.lastPosition = `${place}, ${time}`
        } else {
          this.lastPosition = `${placeEn}, ${timeEn}`
        }
      },

      getLastPosition() {
        positionsApi.fetchLast()
          .then(this.updateLastPositionData)
      },
      isLazyArticle(id) {
        return IS_DESKTOP() ? id < this.articles.length - 7 : id < this.articles.length - 1
      },
    },

    i18n: {
      silentTranslationWarn: true,
      messages: {
        fr: {
          fixWebsite: 'Réparer le site',
          theArticlesOfTheTrip: 'Blog de voyage de Pierre Trollé et Benoît Lefebvre après un tour du monde et d’autres aventures',
          lastPosition: 'Dernière position :',
          title: 'Pierre en voyage',
          lastKnownPosition: 'Dernière position connue :',
        },
        en: {
          fixWebsite: 'Fix the website',
          theArticlesOfTheTrip: 'Travel blog of Pierre Trollé and Benoît Lefebvre after a world trip and other adventures',
          lastPosition: 'Last position:',
          title: 'Discover the world with us!',
          lastKnownPosition: 'Last known position:',
        },
      },
    },
  }

</script>

<style scoped>
.article-list {
  display: flex;
  width: 100%;
  padding: 70px 0 20px;
  justify-content: center;
}

.article-results {
  margin-bottom: 60px;
}

.article-results__title {
  font-weight: 300;
  font-size: 24px;
  margin: 0 0 15px;
}

.hidden {
  display: none;
}

.h2 {
  font-size: 18px;
}

.h3 {
  font-size: 18px;
  font-weight: 600;
}

.article-results__list {
  padding: 0;
  display: flex;
  flex-direction: column;
  justify-content: center;
}

.article-results__item {
  list-style-type: none;
  padding: 0;
  margin: 5px;
}

@media only screen and (min-width: 640px) {
  .article-results__list {
    flex-direction: row;
    flex-wrap: wrap;
  }
}

</style>
