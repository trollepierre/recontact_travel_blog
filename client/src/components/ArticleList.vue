<template>
  <!--todo refaire l'arborescence (job ? + page ?)-->
  <div class="page">
    <main class="page__body">
      <div class="page__container">
        <section class="article-results">
          <h1 class="article-results__title hidden">{{ title }}</h1>
          <p class="article-results__title h1">{{ $t("subtitle") }}</p>
          <p class="article-results__title h2">{{ $t("lastKnownPosition") }}
            <span class="article-results__title h3">{{ lastPosition }}</span></p>
          <template v-if="adminMode">
            <a href="http://recontact.me/apo/sub">
              <button class="article-results__buttons article-results__sync_hidden" type="button"
                      @click.prevent="goToSubscriptions">{{ $t("getSubscribers") }}
              </button>
            </a>
            <button class="article-results__buttons" type="button" :disabled="isClickedSync"
                    @click.prevent="synchronise">{{ $t("getNewArticles") }}</button>
            <button class="article-results__buttons article-results__sync_hidden" type="button" :disabled="isClickedSync"
                    @click.prevent="deleteAll">{{ $t("deleteAllArticles") }}
            </button>
            <button class="article-results__buttons article-results__sync_hidden" type="button" :disabled="isClickedSync"
                    @click.prevent="deleteAndSyncAll">{{ $t("deleteAndSyncAllArticles") }}
            </button>
            <br>

            <form class="article-results__form">
              <p class="article-results__text">
                {{ $t("lastPosition") }}
              </p>
              <label class="article-results__label" for="position-place">{{ $t("place") }}</label>
              <input class="article-results__input article-results__place" id="position-place" placeholder="Paris" v-model="place"/>
              <label class="article-results__label" for="position-time">{{ $t("time") }}</label>
              <input class="article-results__input article-results__time" id="position-time" placeholder="le 1er mai 2018" v-model="time"/>
              <button class="article-results__buttons article-results__action--send" @click="updateLastPosition">{{ $t("confirm") }}</button>
            </form>

          </template>
          <ul class="article-results__list">
            <li v-for="article in articles" class="article-results__item">
              <article-card :article="article" :adminMode="adminMode"></article-card>
            </li>
          </ul>
        </section>
      </div>
    </main>
  </div>
</template>

<script>
  import ArticleCard from './ArticleCard';
  import articlesApi from '@/api/articles';
  import positionsApi from '@/api/positions';
  import syncApi from '@/api/sync';
  import notificationsService from '@/services/notifications';
  import articlesSorter from '@/services/articlesSorter';

  export default {
    name: 'ArticleList',
    components: {
      'article-card': ArticleCard,
    },
    props: ['adminMode'],
    data() {
      return {
        articles: [],
        isClickedSync: false,
        lastPosition: '',
        place: null,
        time: null,
      };
    },
    mounted() {
      this.getArticles();
      this.getLastPosition();
    },
    computed: {
      title() {
        return (this.adminMode) ? this.$t('fixWebsite') : this.$t('theArticlesOfTheTrip');
      },
    },
    methods: {
      getArticles() {
        articlesApi.fetchAll()
          .then((articles) => {
            this.articles = articlesSorter.sortByDropboxId(articles);
          });
      },

      updateLastPositionData({ place, time }) {
        this.lastPosition = `${place}, ${time}`;
      },

      getLastPosition() {
        positionsApi.fetchLast()
          .then(this.updateLastPositionData);
      },

      submit(e) {
        e.preventDefault();
        this.updateLastPosition();
      },

      updateLastPosition() {
        const position = {
          place: this.place,
          time: this.time,
        };
        positionsApi.add(position)
          .then(this.updateLastPositionData);
      },

      disableButton() {
        this.isClickedSync = true;
      },

      enableButton() {
        this.isClickedSync = false;
      },

      synchronise() {
        this.trackEvent();
        this.disableButton();
        notificationsService.information(this, this.$t('syncLaunched'));
        syncApi.launch()
          .then(() => {
            notificationsService.removeInformation(this);
            notificationsService.success(this, this.$t('syncDone'));
          })
          .then(() => this.goToHome())
          .catch((err) => {
            notificationsService.removeInformation(this);
            notificationsService.error(this, `${this.$t('syncError')} ${err}`);
          });
      },

      deleteAll() {
        this.disableButton();
        notificationsService.information(this, this.$t('syncLaunched'));
        articlesApi.deleteAll()
          .then(() => {
            notificationsService.removeInformation(this);
            notificationsService.success(this, this.$t('syncDone'));
          })
          .then(() => this.goToHome())
          .catch((err) => {
            notificationsService.removeInformation(this);
            notificationsService.error(this, `${this.$t('syncError')} ${err}`);
          });
      },

      deleteAndSyncAll() {
        this.disableButton();
        notificationsService.information(this, this.$t('syncLaunched'));
        articlesApi.deleteAndSyncAll()
          .then(() => {
            notificationsService.removeInformation(this);
            notificationsService.success(this, this.$t('syncDone'));
          })
          .then(() => this.goToHome())
          .catch((err) => {
            notificationsService.removeInformation(this);
            notificationsService.error(this, `${this.$t('syncError')} ${err}`);
          });
      },
      goToSubscriptions() {
        this.$router.push('/subscriptions');
      },
      trackEvent() {
        this.$ga.event({
          eventCategory: 'Article List',
          eventAction: 'synchronise',
          eventLabel: 'All articles have been synchronised',
        });
      },

      goToHome() {
        this.enableButton();
        this.$router.push('/');
      },
    },

    i18n: {
      messages: {
        fr: {
          getNewArticles: 'Récupérer les nouveaux articles',
          deleteAllArticles: 'Supprimer tous les articles',
          deleteAndSyncAllArticles: 'Supprimer & synchro tous les articles',
          getSubscribers: 'Récupérer les abonnés de Recontact.me',
          fixWebsite: 'Réparer le site',
          theArticlesOfTheTrip: 'Blog de voyage de Pierre Trollé et Benoît Lefebvre après un tour du monde et d’autres aventures',
          syncLaunched: 'La synchronisation est lancée ! Patientez quelques secondes...',
          syncDone: 'La synchronisation s’est effectuée sans problème !',
          syncError: 'Erreur : Problème durant la synchronisation :',
          place: 'Position :',
          time: 'Date :',
          confirm: 'Envoyer',
          lastPosition: 'Dernière position :',
          subtitle: 'Pierre en voyage',
          lastKnownPosition: 'Dernière position connue :',
        },
        en: {
          getNewArticles: 'Synchronise the new articles',
          deleteAllArticles: 'Delete all articles',
          deleteAndSyncAllArticles: 'Delete and synchronise all articles',
          getSubscribers: 'Get the subscribers',
          fixWebsite: 'Fix the website',
          theArticlesOfTheTrip: 'Travel blog of Pierre Trollé and Benoît Lefebvre after a world trip and other adventures',
          syncLaunched: 'The synchronisation is launched! Please wait...',
          syncDone: 'The synchronisation succeeds!',
          syncError: 'Error during the synchronisation:',
          place: 'Position:',
          time: 'Date:',
          confirm: 'Confirm',
          lastPosition: 'Last position:',
          subtitle: 'Traveling Pierre',
          lastKnownPosition: 'Last known position:',
        },
      },
    },
  };

</script>

<style scoped>
  .page__body {
    display: flex;
    width: 100%;
    padding: 20px 0;
    justify-content: center;
  }

  .article-results {
    margin-bottom: 60px;
  }

  .article-results__title {
    font-weight: 300;
    font-size: 24px;
    margin: 0 0 15px;
  }

  .hidden {
    display: none;
  }

  .h2 {
    font-size: 16px;
  }

  .h3 {
    font-size: 16px;
    font-weight: 600;
  }

  .article-results__list {
    padding: 0;
    display: flex;
    flex-direction: column;
  }

  .article-results__item {
    list-style-type: none;
    padding: 0;
    margin: 5px;
  }

  @media only screen and (min-width: 640px) {
    .article-results__list {
      flex-direction: row;
      flex-wrap: wrap;
    }
  }

  .article-results__input {
    background: #ffffff;
    border: 1px solid cadetblue;
    padding: 15px 10px;
    border-radius: 4px;
    width: 230px;
    margin-bottom: 10px;
    font-weight: 700;
  }

  .article-results__buttons {
    text-transform: uppercase;
    color: #f76252;
    background: #ffffff;
    border: 1px solid #d14800;
    cursor: pointer;
    padding: 15px 30px;
    border-radius: 4px;
    width: 230px;
    margin-bottom: 10px;
    font-weight: 700;
  }

  .article-results__sync_hidden {
    color: #f7b5a9;
  }

  .article-results__buttons:hover {
    background: #d14800;
    color: #ffffff;
  }

  .article-results__buttons:disabled,
  .article__footer button:active {

    background: #BDBDBD;
    border-color: #616161;
    color: #FAFAFA;
    cursor: auto;
  }


</style>
