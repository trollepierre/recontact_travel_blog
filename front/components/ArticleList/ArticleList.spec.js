import VueI18n from 'vue-i18n'

import ArticleList from './ArticleList.vue'
import articlesApi from '../../services/api/articles'
import positionsApi from '../../services/api/positions'
import articlesSorter from '../../services/services/articlesSorter'
import { isCecile } from '../../services'
import ArticleCard from '../ArticleCard/ArticleCard.vue'

jest.mock('../../services')

describe('Component | ArticleList.vue', () => {
  let localVue
  let wrapper
  const article = (dropboxId = 59) => ({
    dropboxId,
    frTitle: 'Le titre',
    enTitle: 'The title',
  })

  const fetchedArticles = [article('92'), article('12')]
  const sortedArticles = [article('12'), article('92')]
  const lastPosition = { place: 'Mexico', time: '1er mai 2019' }

  beforeEach(() => {
    console.warn = jest.fn()

    articlesSorter.sortByDropboxId = jest.fn()
    articlesSorter.sortByDropboxId.mockReturnValue(sortedArticles)
    positionsApi.fetchLast = jest.fn()
    positionsApi.fetchLast.mockResolvedValue(lastPosition)
    articlesApi.fetchAll = jest.fn()
    articlesApi.fetchAll.mockResolvedValue(fetchedArticles)
    localVue = createLocalVue()
    localVue.use(VueI18n)
    const propsData = { adminMode: true }
    wrapper = shallowMount(ArticleList, { localVue, propsData })
  })

  it('should be named "ArticleList"', () => {
    expect(wrapper.name()).toEqual('ArticleList')
  })

  describe('template', () => {
    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should contain lazy when more than 9 articles', async () => {
      // Given
      const articles = [article('9'),
        article('8'), article('7'),
        article('6'), article('5'),
        article('4'), article('3'),
        article('2'), article('1'),
      ]

      // When
      wrapper = await shallowMount(ArticleList, { localVue, data() { return { articles } } })

      expect(wrapper.findAll(ArticleCard).at(7).props().lazy).toEqual(false)
      expect(wrapper.findAll(ArticleCard).at(8).props().lazy).toEqual(true)
    })

    it('should remove last position when cecile website', () => {
      // Given
      isCecile.mockImplementation(() => true)

      // When
      wrapper = shallowMount(ArticleList, { localVue })

      // Then
      expect(wrapper.find('.article-results__title.h3').element).toBeUndefined()
    })
  })

  describe('mounted', () => {
    it('should call articles api to fetch articles', () => {
      expect(articlesApi.fetchAll).toHaveBeenCalledWith()
    })

    it('should save articles from api in data articles', () => {
      expect(wrapper.vm.articles).toEqual(sortedArticles)
    })

    it('should call positions api to fetch last position', () => {
      expect(positionsApi.fetchLast).toHaveBeenCalledWith()
    })

    it('should save last position from api in data last position', () => {
      expect(wrapper.vm.lastPosition).toEqual('Mexico, 1er mai 2019')
    })
  })

  describe('computed property #title', () => {
    it('should return "Réparer le site" when site is in adminMode', () => {
      const propsData = { adminMode: true }

      wrapper = shallowMount(ArticleList, { localVue, propsData })

      expect(wrapper.vm.hiddenTitle).toEqual('fixWebsite')
    })

    it('should return "Les articles du voyage" by default', () => {
      wrapper = shallowMount(ArticleList, { localVue })

      expect(wrapper.vm.hiddenTitle).toEqual('theArticlesOfTheTrip')
    })

    it('should return "mon cadeau de saint val" when site is cecile', () => {
      isCecile.mockImplementation(() => true)

      wrapper = shallowMount(ArticleList, { localVue })

      expect(wrapper.vm.title).toEqual('Mon petit Cadeau de Saint Valentin')
    })

    it('should return "title" by default', () => {
      wrapper = shallowMount(ArticleList, { localVue })

      expect(wrapper.vm.title).toEqual('title')
    })

    it('should return "recharge site" when site is in cecile', () => {
      isCecile.mockImplementation(() => true)

      wrapper = shallowMount(ArticleList, { localVue })

      expect(wrapper.vm.subtitle).toEqual(
        'recharge le site si jamais l’article ne s’est pas chargé',
      )
    })

    it('should return "lastKnownPosition" by default', () => {
      wrapper = shallowMount(ArticleList, { localVue })

      expect(wrapper.vm.subtitle).toEqual('lastKnownPosition')
    })
  })

  describe('methods', () => {
    describe('#updateLastPositionData', () => {
      it('should concat place and time', () => {
        const position = {
          place: 'place',
          time: 'time',
        }

        wrapper.vm.updateLastPositionData(position)

        expect(wrapper.vm.lastPosition).toEqual('place, time')
      })
    })
  })

  describe('locales', () => {
    const languages = Object.keys(ArticleList.i18n.messages)

    it('contains 2 languages', () => {
      expect(languages).toHaveLength(2)
      expect(languages).toEqual(['fr', 'en'])
    })

    describe('each language', () => {
      describe('fr', () => {
        const locales = Object.keys(ArticleList.i18n.messages.fr)

        it('contains 5 locales', () => {
          expect(locales).toHaveLength(5)
          expect(locales).toMatchInlineSnapshot(`
            Array [
              "fixWebsite",
              "theArticlesOfTheTrip",
              "lastPosition",
              "title",
              "lastKnownPosition",
            ]
          `)
        })
      })

      describe('en', () => {
        const locales = Object.keys(ArticleList.i18n.messages.en)

        it('contains 5 locales', () => {
          expect(locales).toHaveLength(5)
          expect(locales).toMatchInlineSnapshot(`
            Array [
              "fixWebsite",
              "theArticlesOfTheTrip",
              "lastPosition",
              "title",
              "lastKnownPosition",
            ]
          `)
        })
      })
    })
  })
})
