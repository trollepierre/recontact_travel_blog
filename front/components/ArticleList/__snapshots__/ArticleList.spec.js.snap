// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component | ArticleList.vue template should match snapshot 1`] = `
<div class="page">
  <app-header-stub></app-header-stub>
  <main class="page__body">
    <div class="page__container">
      <section class="article-results">
        <h1 class="article-results__title hidden">
          Fix the website
        </h1>
        <p class="article-results__title h1">
          Discover the world with us!
        </p>
        <p class="article-results__title h2">
          Last known position:
          <span class="article-results__title h3">
            Mexico, 1er mai 2019
          </span></p>
        <admin-dashboard-stub></admin-dashboard-stub>
        <ul class="article-results__list">
          <li class="article-results__item">
            <article-card-stub adminmode="true" article="[object Object]"></article-card-stub>
          </li>
          <li class="article-results__item">
            <article-card-stub adminmode="true" article="[object Object]"></article-card-stub>
          </li>
        </ul>
      </section>
    </div>
  </main>
</div>
`;
